### PR TITLE
Always remove asm temp file, unless explicitly kept

### LINKED ISSUE #11471: Temporary asm file not deleted
When running the test https://github.com/ocaml/ocaml/blob/trunk/testsuite/tests/tool-ocamlopt-save-ir/start_from_emit.ml , a temporary asm file /tmp/camlasmXXXX.s is not removed.  

This is a mild problem for CI machines (where /tmp gets cluttered by temp files eventually), and may indicate an oversight in the implementation of the `-save-ir-after` and `-stop-after` options.

Cc: @gretay-js

### PR DESCRIPTION
**Pull request by nojb**
Apart from returning the temporary file name, the function `Asmgen.asm_filename` creates the file in question as a side-effect. When the compilation pipeline is stopped before the "emit" phase (with the `-stop-after` option), the (empty) temporary file is left in the filesystem, which can fill up the `/tmp` directory in some circonstances (eg CI).

The solution is to always delete the temporary file unless it is being kept explicitly. A different approach would be to create the temporary file lazily to avoid creating it at all if not used.

Fixes #11471

### CODE REVIEW

**Comment by gasche** [comment]
I find the code strange (both before and after the fix, which itself sounds sensible). After the PR it looks like this:

https://github.com/ocaml/ocaml/blob/f67c55fb776e51beae1cd93e3ab72ef31be01e66/asmcomp/asmgen.ml#L210-L236

Two questions:

1. `create_asm` is defined as `should_emit () && (keep_asm || not binary_backend_available)`. Why do we need to write `not create_asm || not keep_asm`, is this not equivalent to just `not create_asm`? (Note: if `binary_backend_available` is set, then `Asmgen.asm_filename` does not create a temporary file.)

2. The code contains two `remove_file asm_filename`, one in an `exceptionally` branch and one in the non-error case. Could we merge these two calls in a single place, for example as an `always` clause for the outer `try_finally` call?

**Comment by nojb** [comment]
> 1. `create_asm` is defined as `should_emit () && (keep_asm || not binary_backend_available)`. Why do we need to write `not create_asm || not keep_asm`, is this not equivalent to just `not create_asm`? (Note: if `binary_backend_available` is set, then `Asmgen.asm_filename` does not create a temporary file.)

I agree the condition is hard to parse. But to answer the question: when `binary_backend_available` is `false` (this is always the case in the compiler), `keep_asm` is `false` and `should_emit ()` is `true`, then `create_asm` is `true` but the assembly file should still be removed. Said differently, `keep_asm = false` does not imply `create_asm = false`.

> 2. The code contains two `remove_file asm_filename`, one in an `exceptionally` branch and one in the non-error case. Could we merge these two calls in a single place, for example as an `always` clause for the outer `try_finally` call?

That sounds right. Fixed.

**Comment by nojb** [comment]
> > 2. The code contains two `remove_file asm_filename`, one in an `exceptionally` branch and one in the non-error case. Could we merge these two calls in a single place, for example as an `always` clause for the outer `try_finally` call?
> 
> That sounds right. Fixed.

Actually, this needs a little care so as to preserve the assembly file in case of assembler error. In place of an `always` clause, I switched to an auxiliary function instead.

**Review by gasche — APPROVED** [review]
This is tricky code, but I agree that the proposal fixes the issue, and I don't see a good way to simplify it further. Approved.

File: asmcomp/asmgen.ml [hunk 1]
```diff
@@ -266,7 +268,7 @@ type middle_end =
 let asm_filename output_prefix =
     if !keep_asm_file || !Emitaux.binary_backend_available
     then output_prefix ^ ext_asm
-    else Filename.temp_file "camlasm" ext_asm
+    else Filename.temp_file "camlasm" ext_asm (* Note: creates temporary file *)
```
Comment (xavierleroy) [inline asmcomp/asmgen.ml]: The comment (on a call to `Filename.temp_file`) is a bit of a tautology :-)

**Review by xavierleroy — COMMENTED** [review]
Thanks for looking into this very minor issue.  I'll trust you on the solution.  One silly remark below.

File: asmcomp/asmgen.ml (hunk 1 above)
Comment (nojb) [reply asmcomp/asmgen.ml]: Indeed; I removed it.

**Comment by nojb** [comment]
Thanks for the review. I added a comment explaining the code a bit for future reference. Planning to merge once CI passes.
